ml-agents/mlagents/trainers/torch/components/reward_providers/rnd_reward_provider.py

```diff
@@ -0,0 +1,87 @@
+import numpy as np
+from typing import Dict
+from mlagents.torch_utils import torch
+
+from mlagents.trainers.buffer import AgentBuffer
+from mlagents.trainers.torch.components.reward_providers.base_reward_provider import (
+    BaseRewardProvider,
+)
+from mlagents.trainers.settings import RNDSettings
+
+from mlagents_envs.base_env import BehaviorSpec
+from mlagents.trainers.torch.utils import ModelUtils
+from mlagents.trainers.torch.networks import NetworkBody
+from mlagents.trainers.torch.encoders import Normalizer
+from mlagents.trainers.settings import NetworkSettings, EncoderType
+
+
+class RNDRewardProvider(BaseRewardProvider):
+    """
+    https://arxiv.org/pdf/1810.12894.pdf
+    """
+
+    def __init__(self, specs: BehaviorSpec, settings: RNDSettings) -> None:
+        super().__init__(specs, settings)
+        self._ignore_done = True
+        self._random_network = RNDNetwork(specs, settings)
+        self._training_network = RNDNetwork(specs, settings)
+        self._reward_normalizer = Normalizer(1)
+        self.optimizer = torch.optim.Adam(
+            self._training_network.parameters(), lr=settings.learning_rate
+        )
+        self._has_updated_once = False
+
+    def evaluate(self, mini_batch: AgentBuffer) -> np.ndarray:
+        with torch.no_grad():
+            target = self._random_network(mini_batch)
+            prediction = self._training_network(mini_batch)
+            unnormalized_rewards = torch.sum((prediction - target) ** 2, dim=1)
```

Comment:
Just curious - do you know if `(prediction - target) ** 2` gets treated as `x*x` or `pow(x, 2)`? I'd imagine the former is more efficient for evaluating (both it and the gradient).

Reply:
Torch converts x ** 2 to pow(x, 2). The two are equivalent, so we can pick the one we are more comfortable with.